### ISSUE TITLE
[nnc] LoopNest::fuseLoops failed to fuse loops with reduce ops inside

### ISSUE THREAD

**Issue by huiguoo** [pytorch/pytorch#60310]
## 🐛 Bug

`fuseLoops` failed to fuse loops if a reduce op exists. The problem is that for reduce ops, it detected a fake ReadAfterWrite dependence between loop iterations and therefore returned without applying `fuse` transformation. But the detection algorithm is wrong and there is no data dependence between loop iterations.

## To Reproduce

Test to reproduce the behavior:

```
  1 import torch
  2 
  3 class kernel_arena_scope(object):
  4     def __enter__(self):
  5         self.scope = torch._C._te.KernelScope()
  6 
  7     def __exit__(self, typ, val, traceback):
  8         self.scope = None
  9 
 10 with kernel_arena_scope():
 11     f32 = torch._C._te.Dtype.Float
 12     i32 = torch._C._te.Dtype.Int
 13 
 14     M, K, N = 64, 64, 64
 15     Mte, Kte, Nte = [torch._C._te.ExprHandle.int(x) for x in [M, K, N]]
 16     A = torch._C._te.Placeholder('A', f32, [Mte, Kte])
 17     B = torch._C._te.Placeholder('B', f32, [Kte, Nte])
 18 
 19     def get_dim_args(dims):
 20         dim_args = []
 21         for dim in dims:
 22             dim_args.append(torch._C._te.DimArg(dim, 'i' + str(len(dim_args))))
 23         return dim_args
 24 
 25     def compute(dims):
 26         m, n, k = dims[0], dims[1], dims[2]
 27         return A.load([m, k]) * B.load([k, n])
 28 
 29     C = torch._C._te.Reduce('C',
 30                 get_dim_args([Mte, Nte]),
 31                 torch._C._te.Sum(),
 32                 compute,
 33                 get_dim_args([Kte]),
 34                 )
 35     D = torch._C._te.Compute(
 36             'D',
 37             get_dim_args([Mte, Nte]),
 38             lambda m, n: C.load([m, n])
 39             )
 40     stmt = torch._C._te.Block([C.stmt(), D.stmt()])
 41     loopnest = torch._C._te.LoopNest(stmt, [D.buf()])
 42     print(f"original stmt: {loopnest}")
 43     cm, cn, ck = loopnest.get_loops_for(C)
 44     dm, dn = loopnest.get_loops_for(D)
 45     print(cm, dm)                                                                                                                                                             
 46     loopnest.fuse_loops([cm, dm])
 47     print(f"after fuse: {loopnest}")
```

Original stmt:
```
  for (int i0 = 0; i0 < 64; i0++) {
    for (int i1 = 0; i1 < 64; i1++) {
      C[i0, i1] = float(0);
      for (int i0_1 = 0; i0_1 < 64; i0_1++) {
        C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
      }
    }
  }
  for (int i0_2 = 0; i0_2 < 64; i0_2++) {
    for (int i1_1 = 0; i1_1 < 64; i1_1++) {
      D[i0_2, i1_1] = C[i0_2, i1_1];
    }
  }
```
`cm`, `dm` stmts to fuse:
```
for (int i0 = 0; i0 < 64; i0++) {
  for (int i1 = 0; i1 < 64; i1++) {
    C[i0, i1] = float(0);
    for (int i0_1 = 0; i0_1 < 64; i0_1++) {
      C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
    }
  }
}
```

```
for (int i0 = 0; i0 < 64; i0++) {
  for (int i1 = 0; i1 < 64; i1++) {
    D[i0, i1] = C[i0, i1];
  }
}
```

**Comment by navahgar** [comment]
This is caused by how we check for loop-carried dependences after fusing the loops.

Here is the high-level algorithm we currently use to check if two accesses to a buffer, A and B, one of which is a Store, result in a loop-carried dependence:
   1. If the index expressions are equal in A and B, then that is a loop-independent dependence.
   2. If the index expressions are not equal in A and B:
         a) if the bounds on the accesses overlap, then this is a loop-carried dependence.
         b) if the bounds on the accesses do not overlap, then there is no dependence.

The problem with this approach is that, for the merged loop from above:
```
for (int i0 = 0; i0 < 64; i0++) {
  for (int i1 = 0; i1 < 64; i1++) {
    C[i0, i1] = float(0);
    for (int i0_1 = 0; i0_1 < 64; i0_1++) {
      C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0}); // S1
    }
  }
  for (int i2 = 0; i2 < 64; i2++) {
    D[i0, i2] = C[i0, i2];  // S2
  }
}
```
it reports a loop-carried dependence between the statements `S1` and `S2`, because the second dim indices `i1` and `i2` are not "equal", and the accesses overlap (fully in this case).

**Comment by navahgar** [comment]
One approach to solving this is to check if the indices are "equivalent" (in addition to the checks above), as follows:
  * For every pair of indices `Ai`, `Bi` in `A` and `B` that are not "equal":
     * Substitute all loop variables in `Ai` with their `start` bounds. // `Ai_start`.
     * Substitute all loop variables in `Bi` with their `start` bounds. // `Bi_start`.
     * Substitute all loop variables in `Ai` with their `stop` bounds. //  `Ai_stop`.
     * Substitute all loop variables in `Bi` with their `stop` bounds. // `Bi_stop`.
     * If `Ai_start` == `Bi_start` and `Ai_stop` == `Bi_stop`, then there is a loop-independent dependence between A and B.

**Comment by ZolotukhinM** [comment]
I suspect that we can get away with a simpler check.

Let's look at our loops:
```
 for (int i0 = 0; i0 < 64; i0++) {            // L_C0
    for (int i1 = 0; i1 < 64; i1++) {         // L_C1
      C[i0, i1] = float(0);
      for (int i0_1 = 0; i0_1 < 64; i0_1++) { // L_C01
        C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
      }
    }
  }
  for (int i0_2 = 0; i0_2 < 64; i0_2++) {      // L_D0
    for (int i1_1 = 0; i1_1 < 64; i1_1++) {    // L_D1
      D[i0_2, i1_1] = C[i0_2, i1_1];
    }
  }
```

And we want to fuse loops L_C0 and L_D0.

My theory is that we do not need to check indexes of the inner loop because at when one iteration of L_C0 is finished, let's say for `i0=5`, then there would be no further writes to `C[5, *]` and thus it can be safely used in the computation of `D`.
In other words, L_C1 and L_C01 might write to `C[5, *]` in a different order than it would be then read by L_D0 and L_D1, but all these writes will anyway happen before all the reads, so it still is valid.
Practically that would mean that we only need to check indices of the outer loops and ignore indices on inner ones.

Am I missing anything?

**Comment by navahgar** [comment]
> Practically that would mean that we only need to check indices of the outer loops and ignore indices on inner ones.

That would definitely work and is precisely what I had in mind. But the issue is it may not be possible to clearly differentiate between outer loop indices and inner loop indices. The inner loop variables might be used as indices in different dims of buffer accesses, may be even in combination with outer loop variables. For example:
```
  for (int i0 = 0; i0 < 64; i0++) {            // L_C0
    for (int i1 = 0; i1 < 64; i1++) {         // L_C1
      C[i0, i0 + i1] = float(0);
    }
  }
  for (int i0_2 = 0; i0_2 < 64; i0_2++) {      // L_D0
    for (int i1_1 = 0; i1_1 < 64; i1_1++) {    // L_D1
      D[i0_2, i1_1] = C[i0_2, i0_2 + i1_1];
    }
  }
```
In the code above, once we fuse `L_C0` and `L_D0`, we have to check for loop-carried dependence between writes to `C[i0, i0 + i1]` and reads from `C[i0, i0 + i1_1]` (`i0_2` replaced with `i0` here after fusion).

But the expressions `i0 + i1` and `i0 + i1_1` do not match.

**Comment by huiguoo** [comment]
If I understand correctly, Mikhail's approach (only checking the indices that include 'i0') works for this special case, and also works for cases that can not be handled by extending the equivalent check. See the example below,

```
 for (int i0 = 0; i0 < 64; i0++) {            // L_C0
    for (int i1 = 0; i1 < 32; i1++) {         // L_C1
      C[i0, i1] = float(0);
      for (int i0_1 = 0; i0_1 < 64; i0_1++) { // L_C01
        C[i0, i1] = ReduceOp((C[i0, i1]) + (A[i0, i0_1]) * (B[i0_1, i1]), reduce_args={i0});
      }
    }
  }
  for (int i0_2 = 0; i0_2 < 64; i0_2++) {      // L_D0
    for (int i1_1 = 0; i1_1 < 64; i1_1++) {    // L_D1
      D[i0_2, i1_1] = C[i0_2, i1_1];
    }
  }
```
I modified the stop value for ` L_C1`. With this modification, the stop value of  `i1` in `C[i0, i1] ` is different with `i1_1` in  `C[i0_2, i1_1]` and the equivalence check would fail but it is still valid to fuse the two loops.

**Comment by navahgar** [comment]
Good point regarding different stop values preventing fusion. 

> only checking the indices that include 'i0'

this is not possible when the indices do not strictly correspond to one loop var, as shown in my example.

**Comment by ZolotukhinM** [comment]
> In the code above, once we fuse L_C0 and L_D0, we have to check for loop-carried dependence between writes to C[i0, i0 + i1] and reads from C[i0, i0 + i1_1] (i0_2 replaced with i0 here after fusion)
Ah, that makes sense. Yeah, your approach sounds more general then, maybe we can just have an early exit for expressions not depending on outer indexes at all - for those we do not need to bother checking bounds.

**Comment by navahgar** [comment]
So, my current proposal of using `start` and `stop` bounds for all loop vars is not going to work. 

How about this instead? For every loop var inside the loop being fused, we replace them with a dummy var, say `x`, in the index expressions. For all other loop vars, we continue to use the start and stop bounds. The index expressions will be compared after substituting them with these.

Would this work for all cases?

**Comment by bertmaher** [comment]
One case I'm thinking through is a fused transpose:
```
for (i : 0, 64)
  for (j : 0, 64)
    C[i, j] = ...
for (i : 0, 64)
  for (j : 0, 64)
    ... = C[j, i]   // Note the transposition
```

I'm not sure if that introduces some new subtleties.

**Comment by bertmaher** [comment]
I'm pretty sure fusing the outer loops would be invalid (I think?) so we want to make sure our check handles it appropriately.

**Comment by navahgar** [comment]
Yes, fusing outer loops in the case of transpose is definitely invalid. 

Lets see how my proposed check would work here. Here is the loop after fusing the outer `i` loop: (changed `j` in second loop to `k` to make it clear that those are two different vars)
```
for (i : 0, 64)
  for (j : 0, 64)
    C[i, j] = ...
  for (k : 0, 64)
    ... = C[k, i]
```

Start index exprs for `C[i,j]` is `C[0,x]`
Start index exprs for `C[k,i]` is `C[x,0]`

Stop index exprs for `C[i,j]` is `C[64,x]`
Stop index exprs for `C[k,i]` is `C[x,64]`

So, both the start and stop index expressions do not match between the load and store here. We will not be fusing this case.

**Comment by ZolotukhinM** [comment]
I'm not sure I follow, how would replacing with a dummy var help?

In @huiguoo example we would have three new dummy variables (for i1, i0_1 and i1_1). How would that help to compare them?

**Comment by navahgar** [comment]
> In @huiguoo example we would have three new dummy variables (for i1, i0_1 and i1_1). How would that help to compare them?

I meant replacing all the inner loop vars with "one dummy variable", so that they are compare as equal always.

**Comment by navahgar** [comment]
Is there something we can do to eliminate all the inner loop variables from the index expressions?

**Comment by ZolotukhinM** [comment]
How about this example:
```
for (i : 0, 64)
  for (j : 0, 64)
    A[i, j] = 0
for (x : 0, 64)
  for (y : 0, 64)
    B[x, y] = A[x == 32 ? x + 16 : x, y]
```

After fusing:

```
for (i : 0, 64)
  for (j : 0, 64)
    A[i, j] = 0
  for (y : 0, 64)
    B[i, y] = A[i == 32 ? i + 16 : i, y]
```
Start index exprs for A[i,j] is C[0,x]
Start index exprs for A[i == 32 ? i + 16 : i, y] is C[0,x]

Start index exprs for A[i,j] is C[64,x]
Start index exprs for A[i == 32 ? i + 16 : i, y] is C[64,x]

The analysis would say fusion is ok, but in fact it is not (a read from A[48] happens before write after fusion).

**Comment by ZolotukhinM** [comment]
> I meant replacing all the inner loop vars with "one dummy variable", so that they are compare as equal always.

I'm not sure that's gonna help. The inner loop structure might differ arbitrarily and we cannot expect that replacing one of the indices with a dummy variable would help a lot. If we truly want to analyze inner loops, I think we need to perform real bound inference, which is hard. In practice I think it would be sufficient to give up in cases where indexes in loads/stores involve indexes of outer loops and are not trivially equal. I think we should not worry about inner indexes at all - i.e. that's too hard to pull off in practice

**Comment by navahgar** [comment]
Good point @ZolotukhinM. So, using start and stop bounds is not going to cut it. We have to literally do a symbolic execution of all iterations of the outer loop to compare the indices :(

**Comment by navahgar** [comment]
> In practice I think it would be sufficient to give up in cases where indexes in loads/stores involve indexes of outer loops and are not trivially equal. 

This is definitely doable, but we will be missing some cases. Those cases may not come up in our flow and even if they come up we have to live with not fusing it.

**Comment by ZolotukhinM** [comment]
> So, using start and stop bounds is not going to cut it. We have to literally do a symbolic execution of all iterations of the outer loop to compare the indices :(

That's my point. I think handling a general case is impractical here, but we can get away with a simple solution that would cover 99% of the real cases. And the solution would be to require identity in index expressions for indexes containing outer loops variables.

**Comment by navahgar** [comment]
I ll go ahead and implement that. Thanks for the suggestions and discussion.

**Comment by huiguoo** [comment]
And I have a PR to allow users to still use fusion when our correctness checks failed : PR #60601
